### ISSUE TITLE
Neo.DatabaseError.General.UnknownError: Kernel API returned non-existent relationship type: -1

### ISSUE THREAD

**Issue by Arashrohani** [neo4j/neo4j#12140]
I'm getting this error after upgrading to 3.5.1 with the following query:

<pre><code>MATCH (n1:Node1)<-[:Rel1]-()<-[:Rel]-(n2:Node2)
WITH distinct n2, n1
MATCH (n2)<-[Rel:Rel2]-(n3:Node3)
WITH Rel, n3, n2, n1
MATCH (n3)-[:Rel3]->(n4)
WITH n3, n2, Rel, n4, n1
MATCH (n5:Node5)
WHERE (n5.id = "...")
OPTIONAL MATCH (n5)-[:Rel4]->(n6:Node6)
WITH n3, n5, n6, n4, Rel, n2, n1
OPTIONAL MATCH (n5)-[:Rel5*1..]->(n7:Node5)-[:Rel6]->(n7)
WITH distinct n3, n5, n6, n7, n4, Rel, n2, n1
return n3, n5, head(collect(n7)), n4, Rel, n2, n1
</code></pre>

- Neo4j version: 3.5.1
- Operating system: Windows Server 2016
- API/Driver: None; browser client
- **Steps to reproduce**
In empty 3.5.1 database, create a simple structure and run the query above
- Expected behavior: query produces some results.
- Actual behavior: Neo.DatabaseError.General.UnknownError: Kernel API returned non-existent relationship type: -1

Additional Notes: these are the changes that will cause the query to return some records as expected:

Scenario1: n7 node is null in our scenario, when the head(collect()) is removed from return statement.

Scenario2: keep the head(collect()) but remove the Rel from return statement

Scenario3: replace head(collect()) with this Case when n7 IS NULL THEN null else head(collect(n7)) end

this query is working in 3.4 with the same data with no error.

**Comment by felipebn** [comment]
Hi, I'm also having this error message with  Neo4j server 3.5.2-enterpise on Docker.

**Comment by Hunterness** [comment]
Hi, thanks for reporting, we will look into it.

/Team Cypher

**Comment by pontusmelke** [comment]
Hi @Arashrohani and @felipebn 
I have so far not been able to reproduce. Are you running community or enterprise? Could you run the query with EXPLAIN and paste the plan here? Also maybe you could describe how you created your graph?

Do you have more of the error message, maybe you can look in `debug.log` and see if there is a stack trace there? 

Best regards,
Pontus

**Comment by RayaanKhatau** [comment]
Also running into this error when updating to 3.5.2
Will return with more details as I read through the debug log but here is the stack trace I am seeing:
```
java.lang.IllegalStateException: Kernel API returned non-existent relationship type: -1
        at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.getRelationshipTypeById(GraphDatabaseFacade.java:991)
        at org.neo4j.kernel.impl.core.RelationshipProxy.getType(RelationshipProxy.java:219)
        at org.neo4j.kernel.impl.util.RelationshipProxyWrappingValue.type(RelationshipProxyWrappingValue.java:142)
        at org.neo4j.kernel.impl.util.RelationshipProxyWrappingValue.writeTo(RelationshipProxyWrappingValue.java:72)
        at org.neo4j.bolt.v1.messaging.Neo4jPackV1$PackerV1.pack(Neo4jPackV1.java:117)
        at org.neo4j.bolt.v1.messaging.encoder.RecordMessageEncoder.encode(RecordMessageEncoder.java:39)
        at org.neo4j.bolt.v1.messaging.encoder.RecordMessageEncoder.encode(RecordMessageEncoder.java:29)
        at org.neo4j.bolt.v1.messaging.BoltResponseMessageWriterV1.packCompleteMessageOrFail(BoltResponseMessageWriterV1.java:103)
        at org.neo4j.bolt.v1.messaging.BoltResponseMessageWriterV1.write(BoltResponseMessageWriterV1.java:79)
        at org.neo4j.bolt.v1.messaging.ResultHandler$1.visit(ResultHandler.java:47)
        at org.neo4j.bolt.v1.runtime.CypherAdapterStream.lambda$accept$0(CypherAdapterStream.java:75)
        at org.neo4j.cypher.internal.compatibility.v3_5.runtime.executionplan.StandardInternalExecutionResult.accept(StandardInternalExecutionResult.scala:188)
```

**Comment by pontusmelke** [comment]
@RayaanKhatau are you using community or enterprise? If possible please paste a plan, i.e. run the query with `EXPLAIN` prepended.

Regards,
Pontus

**Comment by Arashrohani** [comment]
Hi @pontusmelke, I'm running the enterprise edition and I'm not getting that error after upgrading to 3.5.2, it was happening in 3.5.1 for me. I created a test graph based on the query that I pasted here. maybe those who get the error in 3.5.2 can provide more info. let me know if you still haven't been able to reproduce the error then I may be able to create the graph but you need to run it against the version 3.5.1.
Regards,
Arash

**Comment by pontusmelke** [comment]
That's good to hear @Arashrohani 

I am going to close this issue, there has been one fix that went in 3.5.3 that might cause this. @RayaanKhatau and @felipebn please try out 3.5.3, if there is still an issue please reopen this issue or open a new one. Please provide plan description (just run the query with EXPLAIN), and if you are running community or enterprise, and any other information that will make it easier for us to reproduce the issue.

Best Regards,
Pontus
